src/main/java/org/opensearch/knn/plugin/script/KNNScoringUtil.java

```diff
@@ -101,19 +96,16 @@ public static float l2Squared(List<Number> queryVector, KNNVectorScriptDocValues
      * @return cosine score
      */
     public static float cosinesimilOptimized(float[] queryVector, float[] inputVector, float normQueryVector) {
-        requireEqualDimension(queryVector, inputVector);
-        float dotProduct = 0.0f;
         float normInputVector = 0.0f;
         for (int i = 0; i < queryVector.length; i++) {
-            dotProduct += queryVector[i] * inputVector[i];
             normInputVector += inputVector[i] * inputVector[i];
         }
         float normalizedProduct = normQueryVector * normInputVector;
         if (normalizedProduct == 0) {
             logger.debug("Invalid vectors for cosine. Returning minimum score to put this result to end");
             return 0.0f;
         }
-        return (float) (dotProduct / (Math.sqrt(normalizedProduct)));
+        return (float) (VectorUtil.dotProduct(queryVector, inputVector) / (Math.sqrt(normalizedProduct)));
```

Comment:
if we want to use dotProduct / normalize for calculate cosine. this would do one more iteration as original L108 doing dotProduct. 

PS, i checked `Lucene#DefaultVectorUtilSupport#cosine(float, float)` would do cosine normalize

Reply:
I used that below, I'll see if I can get it to work with the normVector present in this method

Reply:
i see there is `return (float) (sum / Math.sqrt((double) norm1 * (double) norm2));` in `Lucene#DefaultVectorUtilSupport#cosine(float, float)` so we can use it directly in `public static float cosinesimilOptimized` and without using dotProduct

Reply:
Where are you seeing the `DefaultVectorUtilSupport` class?  I've only been able to find `VectorUtil` so far and that class doesn't have a cosine method that takes floats, only float[]

Reply:
> Where are you seeing the DefaultVectorUtilSupport class?

https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/internal/vectorization/DefaultVectorUtilSupport.java#L65

Reply:
```
 java.lang.AssertionError: expected:<3.0> but was:<2.0>
        at __randomizedtesting.SeedInfo.seed([E07AF7412B86E3E1:E1DE4F36CD093211]:0)
        at org.junit.Assert.fail(Assert.java:89)
        at org.junit.Assert.failNotEquals(Assert.java:835)
        at org.junit.Assert.assertEquals(Assert.java:577)
        at org.junit.Assert.assertEquals(Assert.java:701)
        at org.opensearch.knn.plugin.script.KNNScoringSpaceTests.testCosineSimilarity(KNNScoringSpaceTests.java:62)

```

Reply:
After further investigation, the error was caused by a bug in the test.  In addition, the cosinesimilOptimized method is intended to avoid calculating the normalized query vector when using the same query vector over and over again in scripts.  This calculation is done inherently by the `VectorUtil.cosine` method.  Therefore, I think we should leave this method as is, since Lucene does not currently have a method that accomplishes the same goal.

Reply:
@ryanbogan in which cases we are hitting this optimized function? given that Lucene distance calculation uses SIMD if present, and SIMD gives better performance in those cases do we still think we should use this optimized function? This optimized function was good earlier but if the internal calculations are already optimized why we would like to keep this?

Reply:
@navneet1v this function is only used for when the user is using the same query vector repeatedly for searching.  The idea of the function is that the user would normalize the query vector first, then use the optimized method while passing in the normalization as a parameter.

Reply:
Looks like it's also used in KNNScoringSpace as well: https://github.com/opensearch-project/k-NN/blob/main/src/main/java/org/opensearch/knn/plugin/script/KNNScoringSpace.java#L106

Reply:
SIMD would be better as per out older experiments of SIMD. Also, given that lucene lacks that implementation I fine to remove this optimize cosine code for now.

Reply:
I can see multiple versions of `cosinesimil` and `cosineSimilarity`. Lets just move towards 1 where we use Lucene functions to do the distance calculations and remove all others. 

Some are using optimized and some doesn't. Lets just clean things up and move towards 1 implementation.

Reply:
Sure, I'll incorporate that with this PR then

Reply:
IMHO for those that are very serious about performance, they will normalize their data during preprocessing and use inner product directly. So, I think its okay to not change cosine functionality for now and just focus on dot product and l2 for this optimization.